burn-core/src/optim/rmsprop.rs

```diff
@@ -0,0 +1,500 @@
+                _ => grad.clone().mul_scalar(1. - alpha),
+            };
+            let avg = square_avg_state
+                .square_avg
+                .clone()
+                .sub(grad_avg.clone().powf(2.));
+            (grad, square_avg_state, Self { grad_avg, avg })
+        } else {
+            let grad_avg = match centered_state {
+                Some(state) => state.grad_avg,
+                _ => Tensor::zeros(grad.shape()),
+            };
+            let avg = square_avg_state.square_avg.clone();
+            (grad, square_avg_state, Self { grad_avg, avg })
+        }
+    }
+
+    fn to_device(mut self, device: &B::Device) -> Self {
+        self.grad_avg = self.grad_avg.to_device(device);
+        self.avg = self.avg.to_device(device);
+        self
+    }
+}
+
+/// [RMSPropMomentum](RMSPropMomentum) is to store config status for optimizer.
+/// (, which is stored in [optimizer](RMSProp) itself and not passed in during `step()` calculation)
+pub struct RMSPropMomentum {
+    momentum: f32,
+    epsilon: f32,
+}
+
+impl RMSPropMomentum {
+    /// transform [grad](Tensor) and [RMSPropMomentumState] to the next step
+    pub fn transform<B: Backend, const D: usize>(
+        &self,
+        grad: Tensor<B, D>,
+        centered_state: CenteredState<B, D>,
+        momentum_state: Option<RMSPropMomentumState<B, D>>,
+    ) -> (
+        Tensor<B, D>,
+        CenteredState<B, D>,
+        RMSPropMomentumState<B, D>,
+    ) {
+        let grad = grad
+            .clone()
+            .div(centered_state.avg.clone().sqrt().add_scalar(self.epsilon));
+
+        if self.momentum > 0. {
+            let buf = match momentum_state {
+                Some(state) => state
+                    .buf
+                    .clone()
+                    .mul_scalar(self.momentum)
+                    .add(grad.clone()),
+                _ => grad.clone(),
+            };
+            (buf.clone(), centered_state, RMSPropMomentumState { buf })
+        } else {
+            match momentum_state {
+                Some(state) => (grad.clone(), centered_state, state),
+                _ => (
+                    grad.clone(),
+                    centered_state,
+                    RMSPropMomentumState {
+                        buf: Tensor::zeros(grad.shape()),
+                    },
+                ),
+            }
+        }
+    }
+}
+
+/// [RMSPropMomentumState](RMSPropMomentumState) is to store and pass optimizer step params.
+#[derive(Record, Clone, new)]
+pub struct RMSPropMomentumState<B: Backend, const D: usize> {
+    buf: Tensor<B, D>,
+}
+
+impl<B: Backend, const D: usize> RMSPropMomentumState<B, D> {
+    fn to_device(mut self, device: &B::Device) -> Self {
+        self.buf = self.buf.to_device(device);
+        self
+    }
+}
+
+impl<B: Backend> SimpleOptimizer<B> for RMSProp<B> {
+    type State<const D: usize> = RMSPropState<B, D>;
+
+    fn step<const D: usize>(
```

Comment:
Very basic comment, but I would prefer to see this implementation first in the file bellow the struct definitions. Also a lot of functions don't need to be public, since they are only going to be used from the `step` method.

Reply:
Got it. It looks better now. Will push it to the branch later.